File: 50-powx-n/powx-n.py
class Solution:
    def myPow(self, x: float, n: int) -> float:
        
        # helper function
        def quickPow(x: float, n: int) -> float:
            # edge case
            if n == 0:
                return 1
            
            if (n%2 == 0):
                return quickPow(x*x, n//2)
            return x * quickPow(x, n-1)

        # main function
        ans = quickPow(x, abs(n))
        if (n >= 0):
            return ans
        return 1.0/ans
        
        

File: 50-powx-n/powx-n.cpp
class Solution {
public:
    double quickPow(double x, long long n) {
        // edge case
        if (n == 0) return 1;
        
        if (n%2 == 0) {
            double res = quickPow(x, n/2);
            return res * res;
        }
        return x * quickPow(x, n-1);
    }
    double myPow(double x, int n) {
        // do positive side
        double ans = quickPow(x, labs((long long)n));
        if (n >= 0)  return ans;
        return 1.0/ans; // negative exp
    }
};